bigtable/client/grpc_error.h

```diff
@@ -0,0 +1,60 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_GRPC_ERROR_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_GRPC_ERROR_H_
+
+#include <grpc++/grpc++.h>
+#include <stdexcept>
+#include "bigtable/client/version.h"
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Wrap gRPC errors is a C++ exception.
+ *
+ * It is customary, though not required, for C++ libraries and applications
+ * to use classes in the `std::exception` hierarchy to report errors. Meanwhile,
+ * gRPC reports errors using the `grpc::Status` class. This class wraps the
+ * contents of a `grpc::Status` in an exception class.  If the application is
+ * interested in the details of the exception it can catch it and examine all
+ * fields.  If the application developers simply want to log all exceptions they
+ * can catch `std::exception` and log the contents of `what()`.
+ */
+class GRpcError : public std::runtime_error {
+ public:
+  explicit GRpcError(char const* what, grpc::Status const& status);
+
+  //@{
+  /// @name accessors.
+  grpc::StatusCode error_code() const { return error_code_; }
+  std::string const& error_message() const { return error_message_; }
+  std::string const& error_details() const { return error_details_; }
+  //@}
+
+ private:
+  /// Return the what() string given @p status.
+  static std::string CreateWhatString(char const* what,
+                                      grpc::Status const& status);
+
+ private:
+  grpc::StatusCode error_code_;
```

Comment:
In case this exception will reach users drop usage of grpc::StatusCode and change name away from GRpcError.

Reply:
Why?  The error represents a run-time problem, reported by gRPC, that stops the library from completing its work.  Is not like we want (or could if we wanted to) hide the fact that gRPC is the underlying RPC mechanism.